### PR TITLE
[Plugin] Printify/Stripe integration

### PR DESCRIPTION
**Pull request by alis-khadka**
Living Doc: https://github.com/restarone/violet_rails/issues/1562

### Currency
Printify has two currency settings;

- catalog currency: This is the currency in which we can view different product prices in Printify's catalog.
- billing currency : This is the currency in which the shop owner will be charged by Printify.

We can find further details about it over Printify webpage. [LINK 1](https://help.printify.com/hc/en-us/articles/12203640151953-How-can-I-change-the-currency-on-Printify-), [LINK 2](https://help.printify.com/hc/en-us/articles/4483625794577-In-which-currency-can-I-sell-my-products-)

The Printify's products API does not provide any details regarding the `currency` being used. It only returns the price in figures.
Screenshot 2023-05-26 at 4 54 58 PM

So, to overcome this barrier, we would have to manually set the `currency`, which is being used or has been setup in Printify platform, in Violet-Rails end. It will be saved under a `shop` api-resource's properties as below; [TEST-NET LINK](https://violet-test.net/api_namespaces/53/resources/709)

Screenshot 2023-05-26 at 5 12 42 PM

This `currency` information will be used to show the prices in products-listing, product-detail, cart and Stripe's checkout page.

**Note:** For now, we are using `USD` currency over the test-net.

## HOW TO RUN PRINTIFY PLUGIN INSTALER SCRIPT

https://github.com/restarone/violet_rails/pull/1557

### Export variables
 
- cd to violet_rails root

**Required Variables**
- `export SHOP_NAME="Restarone"`
- `export PRINTIFY_API_KEY="your api key"`

**Optional Variables**

- `export STRIPE_SECRET_KEY="your stripe secret key"`
- `export PRODUCTS_PAGE_SLUG="your products page url"`
- `export BUSINESS_NAME="your business name in printify account"`

### Run Installer

**In development**

- `docker-compose run --rm -e SHOP_NAME -e PRINTIFY_API_KEY -e STRIPE_SECRET_KEY -e PRODUCTS_PAGE_SLUG -e BUSINESS_NAME solutions_app rails r ./plugins/installers/printify_stripe_e_commerce.rb`

**In production**

- `rails r ./plugins/installers/printify_stripe_e_commerce.rb`

**In test server with dokku**
- `dokku run -e SHOP_NAME -e PRINTIFY_API_KEY -e STRIPE_SECRET_KEY -e PRODUCTS_PAGE_SLUG -e BUSINESS_NAME <APP_NAME> rails r ./plugins/installers/printify_stripe_e_commerce.rb`

#### SHOP_NAME:

Name of the printify shop you want to sync to violet rails.

Screenshot 2023-05-30 at 5 30 19 PM

<br/>

#### PRINTIFY_API_KEY:
https://github.com/restarone/violet_rails/assets/50227291/22a2853d-1d99-4fc0-8dc8-2a81e31eb67e

<br/>

#### STRIPE_SECRET_KEY:

https://github.com/restarone/violet_rails/assets/50227291/1d92065c-c04c-4a45-a656-46017466429c

<br/>

#### BUSINESS_NAME:
Business name in your printify account.

https://github.com/restarone/violet_rails/assets/50227291/ce67a088-6412-4c02-8f68-1b2c8392e945

<br/>

#### PRODUCTS_PAGE_SLUG:
The url you want for your products list page. Default is 'products'

https://github.com/restarone/violet_rails/assets/50227291/125c9681-84ac-49c6-ab7e-8e6d784abcbe

<br/>

# STORE MANAGEMENT

### How to hide  product from the store:

Screenshot 2023-06-04 at 9 12 04 AM

https://github.com/restarone/violet_rails/assets/50227291/6698994e-1d33-4073-bc6e-6d18cd88a92c

<br/>

### How to change who pays processing fees (Pass On to customer or Absorb Fees)

If pass_processing_fees_to_customer is set to true, the Convenience fee, ie.  the stripe processing fee (2.9% + $0.30) and stripe tax collection charge (0.5% if sales tax enabled) will be passed to customers

Screenshot 2023-06-04 at 9 33 51 AM
Screenshot 2023-06-04 at 9 34 32 AM

https://github.com/restarone/violet_rails/assets/50227291/6b703d35-7b29-437f-b49e-43947cb4e9b1

<br/>

# DEBUGGING

If any feature is not working, please check `shop_logs` namespace. We store error logs there.

### Stuck in `publishing` state?

- Run sync_printify_products plugin

https://github.com/restarone/violet_rails/assets/50227291/08e24d9f-b816-4cb8-9aa5-376f1771e7f1

### How to check if Printify webhooks are created?

- On rails console

```
PRINTIFY_API_KEY="your api key"
```

```
PRINTIFY_HEADERS = {                                             
  "Content-Type" => 'application/json;charset=utf-8',
  "Authorization" => "Bearer #{PRINTIFY_API_KEY}" 
}

shop_id = ApiNamespace.friendly.find('shops').api_resources.first.properties['printify_shop_id']

response = HTTParty.get("https://api.printify.com/v1/shops/#{shop_id}/webhooks.json", headers: PRINTIFY_HEADERS)

JSON.parse(response.body)
```

### CODE REVIEW

**Comment by donrestarone** [comment]
merge this first: https://github.com/restarone/violet_rails/pull/1555

File: installer/printify_plugin.rb [hunk 1]
```diff
@@ -4,9 +4,9 @@
 
 PRODUCTS_PAGE_SLUG = 'products'
 
-API_KEY = "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJhdWQiOiIzN2Q0YmQzMDM1ZmUxMWU5YTgwM2FiN2VlYjNjY2M5NyIsImp0aSI6ImVkNTVlNzlhOGFkMWVkMGJjNDczYjVkYzllMWM1NjE2OTEwMmVhOWY1MTJiZTk3ZGMzYjlhOWZlNzA0ZWY5NWU3MDcwYzgzODY3ZjAxMTNhIiwiaWF0IjoxNjgzNjI5ODk5LjQ2NDk4OSwibmJmIjoxNjgzNjI5ODk5LjQ2NDk5MiwiZXhwIjoxNzE1MjUyMjk5LjQ1NzAxMSwic3ViIjoiMTMyMDI1NzMiLCJzY29wZXMiOlsic2hvcHMubWFuYWdlIiwic2hvcHMucmVhZCIsImNhdGFsb2cucmVhZCIsIm9yZGVycy5yZWFkIiwib3JkZXJzLndyaXRlIiwicHJvZHVjdHMucmVhZCIsInByb2R1Y3RzLndyaXRlIiwid2ViaG9va3MucmVhZCIsIndlYmhvb2tzLndyaXRlIiwidXBsb2Fkcy5yZWFkIiwidXBsb2Fkcy53cml0ZSIsInByaW50X3Byb3ZpZGVycy5yZWFkIl19.AstXZfBRoP0MQkcWhj22anXBzb1XLGzJ4y3BhkVXuxsaYQ5p9Vg8Dqw-UtKZvxInVJfkxdWRgoOvsWCHR8c"
```
Comment (donrestarone) [inline installer/printify_plugin.rb]: @Pralish please make invalid

File: installer/printify_plugin.rb [hunk 2]
```diff
@@ -0,0 +1,3497 @@
+SHOPS_TO_SYNC = [ENV['SHOP_NAME']]
```
Comment (donrestarone) [inline installer/printify_plugin.rb:1]: please move to `plugins/installers/printify_stripe_e_commerce.rb`

Comment (donrestarone) [reply installer/printify_plugin.rb:1]: or `installers/printify_stripe_e_commerce.rb `

**Comment by donrestarone** [comment]
@alis-khadka Im ready to deploy this once CI and final UAT passes-- please let me know

**Comment by alis-khadka** [comment]
> @alis-khadka Im ready to deploy this once CI and final UAT passes-- please let me know

@donrestarone we can move ahead with this PR.

File: app/models/api_resource.rb [hunk 3]
```diff
@@ -141,7 +141,10 @@ def define_associations
   def handle_associations
     ActiveRecord::Base.transaction do
       api_namespace.associations.each do |association|
```
[the remark this answers is not in the document]
Comment (Pralish) [reply app/models/api_resource.rb:143]: @donrestarone, An exception was raised while deleting a namespace with resources with associations.

https://github.com/restarone/violet_rails/pull/1557/files#diff-bf904f4d61b7ac3180b9ed07ab7a57e4e106f1aa46d3a3ebf0bf15a3289757e5R397-R470

File: test/plugins/installers/printify_plugin_test.rb [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+    #####################   ICONS    ##########################
+
+    # cart-icon snippet should exist
+
+    # trash-icon snippet should exist
+
+    # check-icon snippet should exist
+
+    # carousel-control-next-icon snippet should exist
```
[the remark this answers is not in the document]
Comment (Pralish) [reply test/plugins/installers/printify_plugin_test.rb]: @donrestarone , Complete test cases and documentations are still pending. We have only added basic test cases here. We will be working on test cases in a separate PR.

**Comment by donrestarone** [comment]
[deploying](https://github.com/restarone/violet_rails/actions/runs/5135113504/jobs/9266877459) in a one off deployment to [EIDD](https://github.com/restarone/violet_rails/settings/environments/1074079476/edit)

**Comment by donrestarone** [comment]
Branch printify-integration (at b0d165b9bd7f904090bf087eff089e2bd2dacf85) deployed as release 20230601165356 by runner

CC @Ayon95 @Pralish @alis-khadka

**Pull request merged**